charon/src/ast/types.rs

```diff
@@ -353,16 +353,28 @@ type ByteCount = u64;
 /// Maps fields to their offset within the layout.
 #[derive(Debug, Default, Clone, PartialEq, Eq, Serialize, Deserialize, Drive, DriveMut)]
 pub struct VariantLayout {
-    /// The offset of each field. `None` if it is not knowable at this point, either because of
-    /// generics or dynamically-sized types.
+    /// The offset of each field.
     #[drive(skip)]
     pub field_offsets: Vector<FieldId, ByteCount>,
 }
 
+/// Layout of the discriminant with its offset and representation type.
+///
+/// Does not include information about the value range.
+#[derive(Debug, Default, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub struct DiscriminantLayout {
+    /// The offset of the discriminant in bytes.
+    pub offset: ByteCount,
+    /// The representation type of the discriminant.
+    /// If the discriminant is in a niche of a non-scalar type,
+    /// this is `None`.
```

Comment:
```suggestion
    /// If the discriminant is in a niche or a non-scalar type,
    /// this is `None`.
```

Reply:
In fact that looks incorrect: for `NicheAdt` in the tests we do get something. I think supporting niches isn't hard

Reply:
The support for integer niches was intentional (and motivated by `NicheAdt`). Depending on whether we want to support other niches as well (and in general have more information about them in the layout), we could allow more than just integers.

Reply:
Yep no I'm the one confused